packages/extension-url-mapping/src/setup-app.ts

```diff
@@ -0,0 +1,187 @@
+            // Example higher-order component, this can be safely removed.
+            sampleHOC,
+            // Optionally include state for this extension using `withApplicationExtensionStore`
+            (component: React.ComponentType<any>) =>
+                withApplicationExtensionStore(component, {
+                    id: extensionMeta.id,
+                    initializer: sliceInitializer
+                })
+        ]
+
+        return applyHOCs(App, HOCs)
+    }
+
+    /**
+     * TODO: update this comment
+     * This method is used to make changes to the PWA-Kit application routes. If your extension adds a new page to the application
+     * then you can add it to the router here. The routes passed to this method is an accrued list of routes that have been added
+     * from extensions applied before it. It is called during the `getRoutes` phase on both the server and the client.
+     *
+     * NOTE: If you instead want to modify a list of all the routes, refer to the `beforeRouteMatch` below.
+     */
+    async getRoutes({locals}: GetRoutesParams): Promise<RouteProps[]> {
+        if (!locals.originalUrl) {
+            return Promise.resolve([])
+        }
+
+        // Example of API call to get url mapping
+        const config = this.getConfig()
+        const shopperSeo = await getShopperSeoClient(locals, config)
+        const urlMapping = await shopperSeo.getUrlMapping({
+            parameters: {urlSegment: locals.originalUrl}
+        })
+        console.log('--- url mapping', urlMapping)
+
+        const requestURL = new URL(locals.originalUrl, getAppOrigin(locals))
+
+        // Returns a partial route
+        return Promise.resolve([
+            {
+                path: requestURL.pathname,
+                componentName: 'Foo',
+                exact: true
+                // component: SamplePage
+            }
+        ])
+    }
+
+    /**
+     * TODO: update comment to make it clearer that beforeRouteMatch is also called on the client side
+     * This method is used on the server during the rendering pipeline. It's provided a list of all the routes that your application
+     * is configured with, including those defined in the base application and those added by all the extensions. You can use this
+     * method to modify these routes in any way you want, but you must return an array of routes as a result.
+     */
+    beforeRouteMatch({allRoutes, locals}: BeforeRouteMatchParams): RouteProps[] {
+        console.log('--- beforeRouteMatch: initial routes', allRoutes)
+        const index = allRoutes.findIndex(
+            (route) =>
+                // @ts-ignore
+                route.componentName === 'Foo'
+        )
+        if (index === -1) {
+            return allRoutes
+        }
+
+        // Complete the partial route
+        const routes = allRoutes.slice()
+        const [myRoute] = routes.splice(index, 1)
+        myRoute.component = routeComponent(SamplePage, true, locals)
```

Comment:
And then completing the route with the appropriate component. Here the example code is a bit too simplistic. In real life, there won't be hardcoding of the component.. there will be some component mapping logic.